graphql_compiler/compiler/ir_lowering_match.py

```diff
@@ -584,6 +676,282 @@ def prune_output_blocks_in_compound_match_query(compound_match_query):
+    elif isinstance(expression, TernaryConditional):
+        return _simplify_ternary_conditional(expression)
+    elif isinstance(expression, BetweenClause):
+        lower_bound = expression.lower_bound
+        upper_bound = expression.upper_bound
+        if isinstance(lower_bound, ContextField) or isinstance(upper_bound, ContextField):
+            raise AssertionError(u'Found BetweenClause with ContextFields as lower/upper bounds. '
+                                 u'This should never happen: {}'.format(expression))
+    elif isinstance(expression, (OutputContextField, FoldedOutputContextField)):
+        raise AssertionError(u'Found unexpected expression of type {}. This should never happen: '
+                             u'{}'.format(type(expression).__name__, expression))
+    elif isinstance(expression, no_op_blocks):
+        return expression
+    else:
+        raise AssertionError(u'Found unexpected expression of type {}. This should never happen: '
+                             u'{}'.format(type(expression).__name__, expression))
+
+
+def _lower_filters_in_match_traversals(match_traversals, visitor_fn):
+    """Return new match traversals, lowering filters involving non-existent ContextFields.
+
+    Expressions involving non-existent ContextFields are evaluated to True.
+    BinaryCompositions, where one of the operands is lowered to a TrueLiteral,
+    are lowered appropriately based on the present operator (u'||' and u'&&' are affected).
+    TernaryConditionals, where the predicate is lowerd to a TrueLiteral,
+    are replaced by their if_true predicate.
+    The `visitor_fn` implements these behaviors (see `_update_context_field_expression`).
+
+    Args:
+        match_traversals: list of match traversal enities to be lowered
+        visitor_fn: visit_and_update function for lowering expressions in given match traversal
+
+    Returns:
+        new list of match_traversals, with all filter expressions lowered
+    """
+    new_match_traversals = []
+    for match_traversal in match_traversals:
+        new_match_traversal = []
+        for step in match_traversal:
+            if step.where_block is not None:
+                new_filter = step.where_block.visit_and_update_expressions(visitor_fn)
+                if new_filter.predicate == TrueLiteral:
+                    new_filter = None
+
+                new_step = step._replace(where_block=new_filter)
+            else:
+                new_step = step
+
+            new_match_traversal.append(new_step)
+        new_match_traversals.append(new_match_traversal)
+
+    return new_match_traversals
+
+
+def lower_context_field_expressions_in_compound_match_query(compound_match_query):
+    """Lower Expressons involving non-existent ContextFields."""
+    if len(compound_match_query.match_queries) == 0:
+        raise AssertionError(u'Received CompoundMatchQuery with '
+                             u'an empty list of MatchQuery objects.')
```

Comment:
nit: this is claiming it's an empty list, but it only checks the size part and not the list part -- include the actual received object in the error message to guard against accidentally misleading future debuggers of this code